src/Sarif.Multitool/Rules/SARIF2004.OptimizeFileSize.cs

```diff
@@ -69,10 +77,46 @@ private bool HasResultLocationsWithUriAndIndex(string resultPointer)
 
         private bool HasLocationOnlyArtifacts(string artifactPointer)
         {
-            JToken artifactToken = artifactPointer.ToJToken(Context.InputLogToken);
+            var artifactToken = artifactPointer.ToJToken(Context.InputLogToken);
             return
                 artifactToken.HasProperty(SarifPropertyName.Location) &&
                 artifactToken.Children().Count() == 1;
         }
+
+        private void AnalyzeIdOnlyRules(Run run, string runPointer)
+        {
+            // We only verify first item in the rules array,
+            // since tools will typically generate similar nodes.
+            // This approach may cause occasional false negatives.
+
+            ReportingDescriptor firstRule = run?.Tool?.Driver?.Rules?.FirstOrDefault();
```

Comment:
>run?.Tool?.Driver? [](start = 44, length = 18)

You don't need to test `run` because the visitor won't call the `Analyze` method with a null run. And `Tool` and `Driver` are required, so the "schema validation" phase of the validator tool will catch this before we ever get here. So `run.Tool.Driver?.Rules?.FirstOrDefault()` will do.

Reply:
oh correct! right and right, they mandatory!

---
In reply to: [447983469](https://github.com/microsoft/sarif-sdk/pull/1950#discussion_r447983469) [](ancestors = 447983469)

Reply:
note: it should be run.Tool.Driver.Rules?.FirstOrDefault() 

---
In reply to: [447987160](https://github.com/microsoft/sarif-sdk/pull/1950#discussion_r447987160) [](ancestors = 447987160,447983469)